### PR TITLE
Replace browse implementation

### PR DESCRIPTION
**Pull request by roberthein**
img_2063
img_2064
img_2065

### CODE REVIEW

File: Toshi/Controllers/Apps/Controllers/BrowseDelegate.swift [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import Foundation
+import UIKit
+
+class BrowseDelegate: NSObject, UICollectionViewDelegate {
```
Comment (yuliaveres) [inline Toshi/Controllers/Apps/Controllers/BrowseDelegate.swift]: delegate class ?  ? some NSObject which takes control over app navigation ?
i dnt like it. so mixed and complicated

please,
1 - remove this class;
2 - remove datasource class (why u separated them into separate classes?)
3 - leave both responsibilities inside controller, much simpler and easier to follow.

File: Toshi/Controllers/Apps/Controllers/BrowseDataSource.swift [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+            let cell = collectionView.dequeue(BrowseAppCell.self, for: indexPath)
+            cell.item = items[collectionView.section][indexPath.item]
+            
+            return cell
+        }
+        
+        let type = browseTypes[indexPath.item]
+        
+        let cell = collectionView.dequeue(BrowseCell.self, for: indexPath)
+        cell.collectionView.dataSource = collectionView.dataSource
```
Comment (yuliaveres) [inline Toshi/Controllers/Apps/Controllers/BrowseDataSource.swift]: agree that it looks so weird .. 😞

File: Toshi/Controllers/Apps/Controllers/BrowseSearchResultView.swift [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    
+    override var canBecomeFirstResponder: Bool {
+        return true
+    }
+}
+
+extension BrowseSearchResultView: UITableViewDelegate {
+    
+    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        let item = searchResults[indexPath.row]
```
Comment (yuliaveres) [inline Toshi/Controllers/Apps/Controllers/BrowseSearchResultView.swift]: please, double check for indexPath validity

File: Toshi/Controllers/Apps/Controllers/BrowseSearchResultView.swift [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+        Navigator.push(ContactController(contact: item))
+    }
+}
+
+extension BrowseSearchResultView: UITableViewDataSource {
+    
+    func tableView(_: UITableView, heightForRowAt _: IndexPath) -> CGFloat {
+        return UITableViewAutomaticDimension
```
Comment (yuliaveres) [inline Toshi/Controllers/Apps/Controllers/BrowseSearchResultView.swift]: you can omit this, as is by default

File: Toshi/Controllers/Apps/Controllers/BrowseSearchResultView.swift [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    }
+    
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        return searchResults.count
+    }
+    
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = tableView.dequeue(SearchResultCell.self, for: indexPath)
+        cell.app = searchResults[indexPath.row]
```
Comment (yuliaveres) [inline Toshi/Controllers/Apps/Controllers/BrowseSearchResultView.swift]: same here, please don't bind cells with model objects. cell should **JUST** display text / image etc, while models should be owned by one object.
configure it from datasource owner, from here

File: Toshi/Controllers/Apps/Views/AppCell.swift [hunk 6]
```diff
@@ -1,107 +1,107 @@
...
-        self.ratingView.topAnchor.constraint(equalTo: self.categoryLabel.bottomAnchor, constant: 5).isActive = true
-        self.ratingView.centerXAnchor.constraint(equalTo: self.contentView.centerXAnchor).isActive = true
-    }
-
-    required init?(coder _: NSCoder) {
-        fatalError("init(coder:) has not been implemented")
-    }
-}
+//// Copyright (c) 2017 Token Browser, Inc
+////
```
Comment (yuliaveres) [inline Toshi/Controllers/Apps/Views/AppCell.swift]: no no no 😏

File: Toshi/Controllers/Apps/Views/SearchResultCell.swift [hunk 7]
```diff
@@ -19,15 +19,11 @@ class SearchResultCell: UITableViewCell {
     var app: TokenUser? {
         didSet {
```
Comment (yuliaveres) [inline Toshi/Controllers/Apps/Views/SearchResultCell.swift]: should be moved away from here

File: Toshi/Controllers/Apps/Controllers/BrowseAllController.swift [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+
+class BrowseAllController: UITableViewController {
+    
+    var searchResults: [TokenUser] = [] {
+        didSet {
+            tableView.reloadData()
+        }
+    }
+    
+    let type: BrowseType
```
Comment (yuliaveres) [inline Toshi/Controllers/Apps/Controllers/BrowseAllController.swift]: please, give it more explicit naming, here I see it is `BrowseType,` but should I see it later, no idea what is meant by TYPE

File: Toshi/Controllers/Apps/Controllers/BrowseAllController.swift [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+    
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+    
+    init(type: BrowseType) {
```
Comment (yuliaveres) [inline Toshi/Controllers/Apps/Controllers/BrowseAllController.swift]: `init(browseType: BrowseType) {`

File: Toshi/Controllers/Apps/Controllers/BrowseSearchResultView.swift (hunk 3 above)
Comment (roberthein) [inline Toshi/Controllers/Apps/Controllers/BrowseSearchResultView.swift]: Why do you say this? Because you want to prevent `index out of bounds`?

Comment (yuliaveres) [inline Toshi/Controllers/Apps/Controllers/BrowseSearchResultView.swift]: YES

Comment (roberthein) [inline Toshi/Controllers/Apps/Controllers/BrowseSearchResultView.swift]: what I think would be nicer is:
```
if let item = searchResults[safe: indexPath.row] {
    // use item here
}
```
We can achieve this by implementing the following extension:
```
extension Collection {
    subscript (safe index: Index) -> Element? {
        return indices.contains(index) ? self[index] : nil
    }
}
```
Would you agree with me?

Comment (roberthein) [inline Toshi/Controllers/Apps/Controllers/BrowseSearchResultView.swift]: Safe collection access. 🤘

Comment (yuliaveres) [inline Toshi/Controllers/Apps/Controllers/BrowseSearchResultView.swift]: i am very much sure it is much simpler and easier to call count on array then Contains on indices. But as you wish

File: Toshi/APIClients/idAPIClient.swift [hunk 10]
```diff
@@ -331,7 +347,10 @@ public class IDAPIClient: NSObject, CacheExpiryDefault {
             self.teapot.get("/v1/search/user?query=\(name)") { (result: NetworkResult) in
                 switch result {
                 case .success(let json, _):
-                    guard let dictionary = json?.dictionary, let json = dictionary["results"] as? [[String: Any]] else { completion([]); return }
+                    guard let dictionary = json?.dictionary, let json = dictionary["results"] as? [[String: Any]] else {
+                        completion([])
+                        return
```
Comment (Elland) [inline Toshi/APIClients/idAPIClient.swift]: We use a newline before returns for clarity.

File: Toshi/Controllers/Apps/Controllers/BrowseAllController.swift [hunk 11]
```diff
@@ -9,14 +9,14 @@ class BrowseAllController: UITableViewController {
         }
     }
     
-    let type: BrowseType
+    let contentSection: BrowseContentSection
```
Comment (Elland) [inline Toshi/Controllers/Apps/Controllers/BrowseAllController.swift]: `BrowseContentSectionType`, maybe?

File: Toshi/Controllers/Apps/Controllers/BrowseAllController.swift [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+    private func configureTableView() {
+        tableView.separatorStyle = .none
+        tableView.alwaysBounceVertical = true
+        tableView.showsVerticalScrollIndicator = true
+        tableView.contentInset.bottom = 60
+        tableView.register(SearchResultCell.self)
+    }
+    
+    private func fetchData() {
+
```
Comment (yuliaveres) [inline Toshi/Controllers/Apps/Controllers/BrowseAllController.swift]: some of methods have this line in the beginning - some no :/
maybe lets avoid them ? wdyt ?

File: Toshi/Controllers/Apps/Controllers/BrowseAppCell.swift [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+    lazy var name: UILabel = {
+        let label = UILabel(withAutoLayout: true)
+        label.font = Theme.medium(size: 15)
+        label.textColor = Theme.darkTextColor
+        label.numberOfLines = 2
+        
+        return label
+    }()
+    
+    lazy var category: UILabel = {
```
Comment (yuliaveres) [inline Toshi/Controllers/Apps/Controllers/BrowseAppCell.swift]: also, please, name it "categoryLabel"

File: Toshi/Controllers/Apps/Controllers/BrowseAppCell.swift [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+        fatalError("init(coder:) has not been implemented")
+    }
+    
+    override init(frame: CGRect) {
+        super.init(frame: frame)
+        
+        backgroundColor = nil
+        isOpaque = false
+        
+        contentView.addLayoutGuide(verticalPositionGuide)
```
Comment (yuliaveres) [inline Toshi/Controllers/Apps/Controllers/BrowseAppCell.swift]: addSubviews? i didn't mention in every single class, but it should be same

File: Toshi/Controllers/Apps/Controllers/BrowseCell.swift [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+        view.isOpaque = false
+        view.alwaysBounceHorizontal = true
+        view.showsHorizontalScrollIndicator = true
+        view.contentInset = UIEdgeInsets(top: 0, left: 10, bottom: 0, right: 10)
+        view.register(BrowseAppCell.self)
+        
+        return view
+    }()
+    
+    lazy var titleLabel: UILabel = {
```
Comment (yuliaveres) [inline Toshi/Controllers/Apps/Controllers/BrowseCell.swift]: private or private(set) please

File: Toshi/Controllers/Apps/Controllers/BrowseCell.swift [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+        fatalError("init(coder:) has not been implemented")
+    }
+    
+    override init(frame: CGRect) {
+        super.init(frame: frame)
+        
+        backgroundColor = nil
+        isOpaque = false
+        
+        contentView.addSubview(collectionView)
```
Comment (yuliaveres) [inline Toshi/Controllers/Apps/Controllers/BrowseCell.swift]: not fixed / addressed

File: Toshi/Controllers/Apps/Controllers/BrowseAppCell.swift [hunk 17]
```diff
@@ -0,0 +1,100 @@
+import Foundation
+import UIKit
+import SweetUIKit
+
+class BrowseAppCell: UICollectionViewCell {
+    
+    lazy var avatar: AvatarImageView = {
```
Comment (roberthein) [inline Toshi/Controllers/Apps/Controllers/BrowseAppCell.swift]: But `avatar`, `name`, `category` and `rating` are being accessed from the controller now, at your request.

Comment (yuliaveres) [inline Toshi/Controllers/Apps/Controllers/BrowseAppCell.swift]: should be kind of `readonly` , private(set)

File: Toshi/Controllers/Apps/Controllers/BrowseCell.swift [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+        view.textColor = Theme.darkTextColor
+        
+        return view
+    }()
+    
+    lazy var seeAllButton: UIButton = {
+        let view = UIButton()
+        view.titleLabel?.font = Theme.regular(size: 17)
+        view.setTitleColor(Theme.tintColor, for: .normal)
+        view.setTitle("See All", for: .normal)
```
Comment (Elland) [inline Toshi/Controllers/Apps/Controllers/BrowseCell.swift]: @yuliaveres can you paste this for the designers? 👍 It's a good idea.

File: Toshi/Controllers/Apps/Controllers/BrowseAllController.swift (hunk 12 above)
Comment (roberthein) [inline Toshi/Controllers/Apps/Controllers/BrowseAllController.swift]: Sometimes it looks better with a white line sometimes it's better without. We don't need more rules I think.

File: Toshi/Controllers/Apps/Controllers/BrowseController.swift [hunk 19]
```diff
@@ -16,271 +16,123 @@
...
-        self.searchController.delegate = self
-
-        self.searchBar.delegate = self
-        self.searchBar.barTintColor = Theme.viewBackgroundColor
-        self.searchBar.tintColor = Theme.tintColor
-        self.searchBar.placeholder = "Search or type the url"
-
-        let searchField = self.searchBar.value(forKey: "searchField") as? UITextField
+        
+        title = "Browse"
```
Comment (roberthein) [inline Toshi/Controllers/Apps/Controllers/BrowseController.swift]: We should. ASAP.

Comment (Elland) [inline Toshi/Controllers/Apps/Controllers/BrowseController.swift]: Then please do. 💪🏽

File: Toshi/Controllers/Apps/Controllers/BrowseCell.swift [hunk 20]
```diff
@@ -0,0 +1,117 @@
+import Foundation
+import UIKit
+import SweetUIKit
+
+protocol BrowseCellSeeAllDelegate {
+    func seeAll(for contentSection: BrowseContentSection)
+}
+
+class SectionedCollectionView: UICollectionView {
```
Comment (Elland) [inline Toshi/Controllers/Apps/Controllers/BrowseCell.swift]: Wait that is all this class does? What is this being used for? Why is this declared inside the cell file?

File: Toshi/Controllers/Apps/Controllers/BrowseCell.swift [hunk 21]
```diff
@@ -0,0 +1,117 @@
...
+        let view = UIButton()
+        view.titleLabel?.font = Theme.regular(size: 17)
+        view.setTitleColor(Theme.tintColor, for: .normal)
+        view.setTitle("See All", for: .normal)
+        view.addTarget(self, action: #selector(seeAllButtonTapped(_:)), for: .touchUpInside)
+        
+        return view
+    }()
+    
+    private lazy var divider: UIView = {
```
Comment (Elland) [inline Toshi/Controllers/Apps/Controllers/BrowseCell.swift]: Again, `separatorView`.

File: Toshi/Controllers/Apps/Controllers/BrowseSearchResultView.swift [hunk 22]
```diff
@@ -0,0 +1,70 @@
+import Foundation
+import SweetFoundation
+import UIKit
+
+class BrowseSearchResultView: UITableView {
+    
+    var searchResults: [TokenUser] = [] {
+        didSet {
+            reloadData()
```
Comment (Elland) [inline Toshi/Controllers/Apps/Controllers/BrowseSearchResultView.swift]: Same as before.

File: Toshi/Controllers/Apps/Controllers/BrowseSearchResultView.swift [hunk 23]
```diff
@@ -0,0 +1,70 @@
...
+        
+        register(SearchResultCell.self)
+    }
+    
+    override var canBecomeFirstResponder: Bool {
+        return true
+    }
+}
+
+extension BrowseSearchResultView: UITableViewDelegate {
```
Comment (Elland) [inline Toshi/Controllers/Apps/Controllers/BrowseSearchResultView.swift]: Why is the view its own data source?

File: Toshi/Views/AvatarImageView.swift [hunk 24]
```diff
@@ -17,47 +17,40 @@ import UIKit
...
-        get {
-            return self.layer.cornerRadius
-        }
-    }
 
     override init(frame: CGRect) {
         super.init(frame: frame)
 
-        self.setup()
+        setup()
```
Comment (Elland) [inline Toshi/Views/AvatarImageView.swift]: It's one thing not being bothered to write it, but why are you actively removing it here?

**Comment by Elland** [comment]
As usual, some comments are marked as outdated, even tho they weren't addressed. Keep an eye out for this. 👍

File: Toshi/APIClients/idAPIClient.swift [hunk 25]
```diff
@@ -231,11 +234,15 @@ public class IDAPIClient: NSObject, CacheExpiryDefault {
                 switch result {
                 case .success(let json, _):
                     // we know it's a dictionary for this API
```
Comment (marijnschilling) [inline Toshi/APIClients/idAPIClient.swift]: I don't understand these comments, maybe it should be removed or it should be explained better.

File: Toshi/APIClients/idAPIClient.swift [hunk 26]
```diff
@@ -256,7 +263,10 @@ public class IDAPIClient: NSObject, CacheExpiryDefault {
...
```
Comment (marijnschilling) [inline Toshi/APIClients/idAPIClient.swift]: Same here, don't understand these comments

File: Toshi/Controllers/Apps/Controllers/BrowseCell.swift [hunk 27]
```diff
@@ -0,0 +1,129 @@
+import Foundation
+import UIKit
+import SweetUIKit
+
+protocol BrowseCellSelectionDelegate {
+    func seeAll(for contentSection: BrowseContentSection)
```
Comment (marijnschilling) [inline Toshi/Controllers/Apps/Controllers/BrowseCell.swift]: Please add a parameter for the sender here as well
`func seeAll(for contentSection: BrowseContentSection, onCollectionView collectionView: SectionedCollectionView)`

File: Toshi/Controllers/Apps/Controllers/BrowseCell.swift [hunk 28]
```diff
@@ -0,0 +1,129 @@
...
+        view.textColor = Theme.darkTextColor
+        
+        return view
+    }()
+    
+    private lazy var seeAllButton: UIButton = {
+        let view = UIButton()
+        view.titleLabel?.font = Theme.regular(size: 17)
+        view.setTitleColor(Theme.tintColor, for: .normal)
+        view.setTitle("See All", for: .normal)
```
Comment (marijnschilling) [inline Toshi/Controllers/Apps/Controllers/BrowseCell.swift]: I always use localizedStrings for every copy in the app, I don't see that happening in Toshi. Is there a reason why we don't? Why do I even do it? 🤔

File: Toshi/Controllers/Apps/Controllers/BrowseCell.swift [hunk 29]
```diff
@@ -0,0 +1,129 @@
+import Foundation
+import UIKit
+import SweetUIKit
+
+protocol BrowseCellSelectionDelegate {
+    func seeAll(for contentSection: BrowseContentSection)
+    func didSelectItem(at indexPath: IndexPath, collectionView: SectionedCollectionView)
```
Comment (marijnschilling) [inline Toshi/Controllers/Apps/Controllers/BrowseCell.swift]: please rename to `onCollectionView` that reads nicer ✌️

File: Toshi/Controllers/Apps/Controllers/BrowseCell.swift (hunk 28 above)
Comment (roberthein) [inline Toshi/Controllers/Apps/Controllers/BrowseCell.swift]: Just added localization.

File: Toshi/Controllers/Apps/Controllers/BrowseCell.swift (hunk 29 above)
Comment (roberthein) [inline Toshi/Controllers/Apps/Controllers/BrowseCell.swift]: Not sure about `onCollectionView:`, maybe it's more swifty to do `on collectionView:`

File: Toshi/Controllers/Apps/Controllers/BrowseCell.swift (hunk 27 above)
Comment (roberthein) [inline Toshi/Controllers/Apps/Controllers/BrowseCell.swift]: Why should there be a parameter that will never be used? Also swiftformat will remove it if it is not being used.

File: Toshi/APIClients/idAPIClient.swift (hunk 25 above)
Comment (roberthein) [inline Toshi/APIClients/idAPIClient.swift]: Me neither, but they are not part of this pr.

File: Toshi/APIClients/idAPIClient.swift (hunk 26 above)
Comment (roberthein) [inline Toshi/APIClients/idAPIClient.swift]: Me neither, but they are not part of this pr.

**Comment by roberthein** [comment]
@marijnschilling 

`In sketch i don’t see the Username added to the cells, only the display name`
I will check this with the design and Android team.
`In sketch I see the Top rated apps on top and below that the Latests apps, in the app it’s now the other way around`
The Android app was leading on this, so it's correct the way it is.
`The search thing on top feels a bit weird, is this a component we want to improve?`
Yes it is. And I agree it could be better.
`Also its just lightly different form the favourites search bar. I think they should at least be the same.`
`Favourites table view has a light grey background, that already looks better in combination with dragging down the search bar.`
I will make them the same. 👍 
`Sketch design say’s “more’ instead of ‘See all’, just checking wether thats correct 😬`
Daniel is checking this.

`When selecting ‘See all’ and going back to browse and when selecting a profile and going back to browse sometimes the contentOffset get’s reset, and sometimes not (alternating) weird! Can you check that out?`
I don't experience this, how do you reproduce this?

File: Toshi/Controllers/Apps/Controllers/BrowseCell.swift (hunk 27 above)
Comment (marijnschilling) [inline Toshi/Controllers/Apps/Controllers/BrowseCell.swift]: Yeah good one, I do it because of habit and its nice so i can recognise delegate methods easier. But I guess I should just stop doing that ✌️

File: Toshi/Controllers/Apps/Controllers/BrowseAppCell.swift [hunk 30]
```diff
@@ -0,0 +1,74 @@
+import Foundation
+import UIKit
+import SweetUIKit
+
+class BrowseAppCell: UICollectionViewCell {
+    
+    private(set) lazy var avatar: AvatarImageView = AvatarImageView()
+    
+    private(set) lazy var name: UILabel = {
```
Comment (yuliaveres) [inline Toshi/Controllers/Apps/Controllers/BrowseAppCell.swift]: i would ask you to rename to `nameLabel,` 
`name` sounds like string
should look weird: `name.text = "..`

Comment (yuliaveres) [inline Toshi/Controllers/Apps/Controllers/BrowseAppCell.swift]: this counts for all the rest ui elements, 🙏

File: Toshi/Controllers/Apps/Controllers/BrowseAppCell.swift [hunk 31]
```diff
@@ -0,0 +1,74 @@
...
+    override init(frame: CGRect) {
+        super.init(frame: frame)
+        
+        backgroundColor = nil
+        isOpaque = false
+        
+        addSubviewsAndConstraints()
+    }
+    
+    func addSubviewsAndConstraints() {
```
Comment (yuliaveres) [inline Toshi/Controllers/Apps/Controllers/BrowseAppCell.swift]: private please

File: Toshi/Controllers/Apps/Controllers/BrowseAllController.swift [hunk 32]
```diff
@@ -0,0 +1,113 @@
...
+        tableView.showsVerticalScrollIndicator = true
+        tableView.contentInset.bottom = 60
+        tableView.register(SearchResultCell.self)
+    }
+    
+    private func fetchData() {
+        
+        switch contentSection {
+        case .topRatedApps:
+            AppsAPIClient.shared.getTopRatedApps(limit: 100, completion: { apps, error in
```
Comment (marijnschilling) [inline Toshi/Controllers/Apps/Controllers/BrowseAllController.swift]: can you change this to trailing closure in al switch cases?
```
AppsAPIClient.shared.getTopRatedApps(limit: 100) { apps, error in
```

File: Toshi/Controllers/Apps/Controllers/BrowseAllController.swift [hunk 33]
```diff
@@ -0,0 +1,113 @@
...
+    }
+    
+    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        if let contact = searchResults.element(at: indexPath.row) {
+            Navigator.push(ContactController(contact: contact))
+        }
+    }
+    
+    override func tableView(_: UITableView, estimatedHeightForRowAt _: IndexPath) -> CGFloat {
+        return UITableViewAutomaticDimension
```
Comment (marijnschilling) [inline Toshi/Controllers/Apps/Controllers/BrowseAllController.swift]: If that is so might be smart to remove it, makes the code easier!

File: Toshi/Controllers/Apps/Controllers/BrowseCell.swift [hunk 34]
```diff
@@ -0,0 +1,131 @@
...
+        super.prepareForReuse()
+        
+        titleLabel.text = nil
+    }
+}
+
+extension BrowseCell: UICollectionViewDelegate {
+    
+    func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
+
```
Comment (marijnschilling) [inline Toshi/Controllers/Apps/Controllers/BrowseCell.swift]: No newline necessary right?

File: Toshi/Controllers/Apps/Controllers/BrowseAppCell.swift (hunk 30 above)
Comment (roberthein) [inline Toshi/Controllers/Apps/Controllers/BrowseAppCell.swift]: done

File: Toshi/Controllers/Apps/Controllers/BrowseAppCell.swift (hunk 31 above)
Comment (roberthein) [inline Toshi/Controllers/Apps/Controllers/BrowseAppCell.swift]: done

File: Toshi/Controllers/Apps/Controllers/BrowseCell.swift [hunk 35]
```diff
@@ -0,0 +1,117 @@
...
+        seeAllButton.right(to: contentView, offset: -15)
+        seeAllButton.height(44)
+        
+        divider.height(Theme.borderHeight)
+        divider.left(to: self, offset: 15)
+        divider.right(to: self)
+        divider.bottom(to: contentView)
+    }
+    
+    func seeAllButtonTapped(_ button: UIButton) {
```
Comment (marijnschilling) [inline Toshi/Controllers/Apps/Controllers/BrowseCell.swift]: I also would like it if we follow the apple style for this ✌️

File: Toshi/Controllers/Apps/Controllers/BrowseCell.swift [hunk 36]
```diff
@@ -0,0 +1,131 @@
+import Foundation
```
Comment (marijnschilling) [inline Toshi/Controllers/Apps/Controllers/BrowseCell.swift]: Do we need to add that header with the `// Copyright (c) 2017 Token Browser, Inc` everywhere? Some classes are missing it.

Comment (roberthein) [inline Toshi/Controllers/Apps/Controllers/BrowseCell.swift]: Added it!

File: Toshi/Controllers/Apps/Controllers/BrowseCell.swift (hunk 34 above)
Comment (roberthein) [inline Toshi/Controllers/Apps/Controllers/BrowseCell.swift]: I think this is fine.

File: Toshi/APIClients/AppsAPIClient.swift [hunk 37]
```diff
@@ -28,27 +28,51 @@ public class AppsAPIClient: NSObject, CacheExpiryDefault {
...
+        }
+    }
+    
+    func getFeaturedApps(limit: Int = 10, completion: @escaping (_ apps: [TokenUser]?, _ error: Error?) -> Void) {
+        DispatchQueue.global(qos: .userInitiated).async {
+            self.teapot.get("/v1/search/apps?top=false&recent=true&limit=\(limit)") { (result: NetworkResult) in
+                switch result {
+                    
+                case .success(let json, _):
+                    guard let json = json?.dictionary else { completion(nil, nil)
```
Comment (roberthein) [inline Toshi/APIClients/AppsAPIClient.swift]: I think it would be better to make a `clean-up api's` pr to do all these things.

File: Toshi/Controllers/Apps/Controllers/BrowseCell.swift (hunk 34 above)
Comment (marijnschilling) [inline Toshi/Controllers/Apps/Controllers/BrowseCell.swift]: No, this is our guidelines we do not add this unnecesary white space🤷‍♀️ 
> Whitespace within methods should separate functionality

File: Toshi/Controllers/Apps/Controllers/BrowseAllController.swift [hunk 38]
```diff
@@ -57,43 +57,43 @@ class BrowseAllController: UITableViewController {
         
         switch contentSection {
         case .topRatedApps:
-            AppsAPIClient.shared.getTopRatedApps(limit: 100, completion: { apps, error in
+            AppsAPIClient.shared.getTopRatedApps(limit: 100) { apps, error in
```
Comment (marijnschilling) [inline Toshi/Controllers/Apps/Controllers/BrowseAllController.swift]: 👯 !

File: Toshi/Controllers/Apps/Controllers/BrowseCell.swift (hunk 34 above)
Comment (marijnschilling) [inline Toshi/Controllers/Apps/Controllers/BrowseCell.swift]: But maybe better to keep an eye on this in future code 🚀  I will approve! for now...
